test/test_slots.py

```diff
@@ -287,3 +263,163 @@ def test_nb_ops_power(self):
         tmp = p
         tmp **= 42
         assert tmp == (p, 'inplace_power', 42, None)
+
+
+class TestSqSlots(HPyTest):
+
+    ExtensionTemplate = PointTemplate
+
+    def test_sq_item_and_sq_length(self):
+        mod = self.make_module("""
+            @DEFINE_PointObject
+
+            HPyDef_SLOT(Point_getitem, Point_getitem_impl, HPy_sq_item);
+            static HPy Point_getitem_impl(HPyContext ctx, HPy self, HPy_ssize_t idx)
+            {
+                return HPyLong_FromLong(ctx, (long)idx*2);
+            }
+
+            HPyDef_SLOT(Point_length, Point_length_impl, HPy_sq_length);
+            static HPy_ssize_t Point_length_impl(HPyContext ctx, HPy self)
+            {
+                return 1234;
+            }
+
+            @EXPORT_POINT_TYPE(&Point_getitem, &Point_length)
+            @INIT
+        """)
+        p = mod.Point()
+        assert p[4] == 8
+        assert p[21] == 42
+        assert len(p) == 1234
+
+    def test_sq_ass_item(self):
+        import pytest
+        mod = self.make_module("""
+            @DEFINE_PointObject
+            @DEFINE_Point_xy
+
+            HPyDef_SLOT(Point_setitem, Point_setitem_impl, HPy_sq_ass_item);
+            static int Point_setitem_impl(HPyContext ctx, HPy self, HPy_ssize_t idx,
+                                          HPy h_value)
+            {
+                PointObject *point = HPy_CAST(ctx, PointObject, self);
+                long value = HPyLong_AsLong(ctx, h_value);
+                if (HPyErr_Occurred(ctx))
+                    return -1;
+                if (idx == 0)
+                    point->x = value;
+                else if (idx == 1)
+                    point->y = value;
+                else {
+                    HPyErr_SetString(ctx, ctx->h_IndexError, "invalid index");
+                    return -1;
+                }
+                return 0;
+            }
+
+            @EXPORT_POINT_TYPE(&Point_x, &Point_y, &Point_setitem)
+            @INIT
+        """)
+        p = mod.Point()
+        p[0] = 100
+        assert p.x == 100
+        p[1] = 200
+        assert p.y == 200
+        with pytest.raises(IndexError):
+            p[2] = 300
```

Comment:
Should we also test what happens with `del p[i]` here?

Reply:
good point, added by 4208706